File: AKAuth0TestApp/AKAuth0TestApp/ViewController.m
//
//  ViewController.m
//  AKAuth0TestApp
//

#import "ViewController.h"
#import "AppDelegate.h"
#import <Lock/Lock.h>

//Please use your Auth0 APIv2 token from https://auth0.com/docs/api/management/v2/tokens
static NSString *kAuth0APIv2Token = @"Auth0APIv2Token";

//Please use your application data from https://manage.auth0.com/#/account/advanced
//section "Global Client Information"
static NSString *kGlobalClientId = @"GlobalClientId";
static NSString *kGlobalClientSecret = @"GlobalClientSecret";

//Please use your application data from https://auth0.com/docs/api/authentication
static NSString *kAuth0ClientId = @"Auth0ClientId";
static NSString *kAuth0ClientSecret = @"Auth0ClientSecret";
static NSString *kAuth0Domain = @"Auth0Domain";

static NSString *kAuth0ConnectionType = @"Username-Password-Authentication";
static NSString *kOpenURLProperty = @"openURL";
static NSString *kAccessToken = @"access_token";

@interface ViewController ()

@property (strong, nonatomic) A0UserProfile *profile;
@property (strong, nonatomic) NSMutableArray *pickerData;
@property (strong, nonatomic) NSMutableArray *userList;
@property (copy, nonatomic) NSString *selectedUser;

@property (weak, nonatomic) IBOutlet UITextField *emailTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;
- (IBAction)clickLoginButton:(id)sender;
@property (weak, nonatomic) IBOutlet UILabel *userName;
@property (weak, nonatomic) IBOutlet UILabel *userId;
@property (weak, nonatomic) IBOutlet UILabel *userEmail;
- (IBAction)clickGetUserList:(id)sender;
@property (weak, nonatomic) IBOutlet UIPickerView *usersPickerView;
- (IBAction)clickImpersonateButton:(id)sender;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.usersPickerView.dataSource = self;
    self.usersPickerView.delegate = self;
    
    self.pickerData = [NSMutableArray arrayWithCapacity:10];
}

// Add observing for "openURL" property of AppDelegate
- (void)viewWillAppear:(BOOL)animated {
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    NSKeyValueObservingOptions kvoOptions = NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew;
    [appDelegate addObserver:self forKeyPath:kOpenURLProperty options:kvoOptions context:nil];
}

- (void)viewDidDisappear:(BOOL)animated {
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    [appDelegate removeObserver:self forKeyPath:kOpenURLProperty];
}

- (void)observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    NSURL *oldURL = change[NSKeyValueChangeOldKey];
    NSURL *newURL = change[NSKeyValueChangeNewKey];
    
    if ([keyPath isEqualToString:kOpenURLProperty] && newURL && ![oldURL isEqual:newURL]) {
        [self loginAsWithURL:newURL];
    }
}

// Step 1: Login to Auth0
- (IBAction)clickLoginButton:(id)sender {
    if (self.emailTextField.text.length < 1) {
        [self showMessage:@"You need to eneter email"];
        return;
    }
    if (self.passwordTextField.text.length < 1) {
        [self showMessage:@"You need to eneter password"];
        return;
    }
    
    NSString *email = self.emailTextField.text;
    NSString *password = self.passwordTextField.text;
    A0APIClient *client = [[A0Lock sharedLock] apiClient];
    A0APIClientAuthenticationSuccess success = ^(A0UserProfile *profile, A0Token *token) {
        self.profile = profile;
    };
    A0APIClientError error = ^(NSError *error){
        NSLog(@"Oops something went wrong: %@", error);
    };
    A0AuthParameters *params = [A0AuthParameters newDefaultParams];
    params[A0ParameterConnection] = kAuth0ConnectionType; // Or your configured DB connection
    [client loginWithUsername:email
                     password:password
                   parameters:params
                      success:success
                      failure:error];
}

// Step 2: Get list of available users for impersonation
- (IBAction)clickGetUserList:(id)sender {
    // GET request
    // We need url "https://<Auth0 Domain>//api/v2/users?include_totals=true&include_fields=true&search_engine=v2"
    // and header "Authorization : Bearer <kAuth0APIv2Token>"

    NSString *apiToken = [NSBundle mainBundle].infoDictionary[kAuth0APIv2Token];
    NSString *bearerToken = [NSString stringWithFormat:@"Bearer %@", apiToken];
    NSDictionary *headers = @{ @"Authorization": bearerToken };
    
    NSString *domain = [NSBundle mainBundle].infoDictionary[kAuth0Domain];
    NSString *urlString = [NSString stringWithFormat:@"https://%@/api/v2/users?include_totals=true&include_fields=true&search_engine=v2", domain];
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:10.0];
    [request setHTTPMethod:@"GET"];
    [request setAllHTTPHeaderFields:headers];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                    if (error) {
                                                        NSLog(@"%@", error);
                                                    } else {
                                                        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                                                        NSLog(@"%@", httpResponse);
                                                        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
                                                        NSLog(@"%@", dict);
                                                        [self createUserList:dict];
                                                    }
                                                }];
    [dataTask resume];
}

// Step 3: Get bearer token for an impersonation
- (IBAction)clickImpersonateButton:(id)sender {
    // POST request
    // We need url "https://<Auth0 Domain>/oauth/token"
    // and header "content-type": "application/json"
    NSString *clientId = [NSBundle mainBundle].infoDictionary[kGlobalClientId];
    NSString *clientSecret = [NSBundle mainBundle].infoDictionary[kGlobalClientSecret];
    NSDictionary *headers = @{ @"content-type": @"application/json" };
    NSDictionary *body = @{ @"client_id": clientId,
                            @"client_secret" : clientSecret,
                            @"grant_type" : @"client_credentials"
                            };
    
    
    NSError *error;
    NSData *dataFromDict = [NSJSONSerialization dataWithJSONObject:body
                                                           options:0
                                                             error:&error];
    
    NSString *domain = [NSBundle mainBundle].infoDictionary[kAuth0Domain];
    NSString *urlString = [NSString stringWithFormat:@"https://%@/oauth/token", domain];
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:10.0];
    [request setHTTPMethod:@"POST"];
    [request setAllHTTPHeaderFields:headers];
    [request setHTTPBody:dataFromDict];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                    if (error) {
                                                        NSLog(@"%@", error);
                                                    } else {
                                                        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
                                                        NSLog(@"%@", dict);
                                                        //Impersonate user
                                                        [self impersonateUserWithBearerToken:dict[kAccessToken]];
                                                    }
                                                }];
    [dataTask resume];
    
}

// Step 4: Perform the impersonation for selected user (user_id) and get a link
- (void)impersonateUserWithBearerToken:(NSString *)bearerToken {
    // POST request
    // We need url "https://<Auth0 Domain>/users/<Selected_User_id>/impersonate"
    // and header "authorization": "Bearer <Bearer_Token>"
    
    NSString *token = [NSString stringWithFormat:@"Bearer %@", bearerToken];
    NSDictionary *headers = @{ @"content-type": @"application/json",
                               @"Authorization": token};
    NSString *clientId = [NSBundle mainBundle].infoDictionary[kAuth0ClientId];
    NSDictionary *body = @{ @"protocol": @"oauth2",
                            @"impersonator_id" : self.profile.userId,
                            @"client_id" : clientId,
                            @"response_type" : @"code"
                            };
    
    NSError *error;
    NSData *dataFromDict = [NSJSONSerialization dataWithJSONObject:body
                                                           options:0
                                                             error:&error];
    
    NSString *userId = [self.selectedUser stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet alphanumericCharacterSet]];
    NSString *domain = [NSBundle mainBundle].infoDictionary[kAuth0Domain];
    NSString *urlString = [NSString stringWithFormat:@"https://%@/users/%@/impersonate", domain, userId];
    NSURL *url = [NSURL URLWithString:urlString];
    
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:10.0];
    [request setHTTPMethod:@"POST"];
    [request setAllHTTPHeaderFields:headers];
    [request setHTTPBody:dataFromDict];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                    if (error) {
                                                        NSLog(@"%@", error);
                                                    } else {
                                                        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *) response;
                                                        NSLog(@"%@", httpResponse);
                                                        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
                                                        NSLog(@"%@", dict);
                                                        NSString *url = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
                                                        NSLog(@"URL : %@", url);
                                                        
                                                        [self openURLImpersonation:url];
                                                    }
                                                }];
    [dataTask resume];
}

// Step 5: Open the link for impersonation
- (void) openURLImpersonation:(NSString *)urlImpersonation {
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:urlImpersonation]];
}

// Step 6: Fetch a token with the code
- (void)fetchTokenWithCode:(NSString *)code callbackURL:(NSString *)redirectURL {
    // POST request
    // We need url "https://<Auth0 Domain>/oauth/token"
    // and header "content-type": "application/json"

    NSString *clientId = [NSBundle mainBundle].infoDictionary[kAuth0ClientId];
    NSString *clientSecret = [NSBundle mainBundle].infoDictionary[kAuth0ClientSecret];
    NSDictionary *headers = @{ @"content-type": @"application/json" };
    NSDictionary *body = @{ @"client_id": clientId,
                            @"client_secret" : clientSecret,
                            @"grant_type" : @"authorization_code",
                            @"redirect_uri" : redirectURL,
                            @"code" : code
                            };
    
    
    NSError *error;
    NSData *dataFromDict = [NSJSONSerialization dataWithJSONObject:body
                                                           options:0
                                                             error:&error];
    
    NSString *domain = [NSBundle mainBundle].infoDictionary[kAuth0Domain];
    NSString *urlString = [NSString stringWithFormat:@"https://%@/oauth/token", domain];
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:10.0];
    [request setHTTPMethod:@"POST"];
    [request setAllHTTPHeaderFields:headers];
    [request setHTTPBody:dataFromDict];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                    if (error) {
                                                        NSLog(@"%@", error);
                                                    } else {
                                                        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
                                                        
                                                        [self fetchUserProfileWithAccessToken:dict[kAccessToken]];
                                                    }
                                                }];
    [dataTask resume];
}

// Step 7: Get user profile for a new user
- (void)fetchUserProfileWithAccessToken:(NSString *)accessToken {
    // GET request
    // We need url "https://<Auth0 Domain>/userinfo/?access_token=<ACCESS_TOKEN>"
    
    NSString *domain = [NSBundle mainBundle].infoDictionary[kAuth0Domain];
    NSString *urlString = [NSString stringWithFormat:@"https://%@/userinfo/?access_token=%@",domain, accessToken];
    NSURL *url = [NSURL URLWithString:urlString];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url
                                                           cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                       timeoutInterval:10.0];
    [request setHTTPMethod:@"GET"];
    
    NSURLSession *session = [NSURLSession sharedSession];
    NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                    if (error) {
                                                        NSLog(@"%@", error);
                                                    } else {
                                                        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:&error];
                                                        NSLog(@"%@", dict);
                                                        self.profile = [[A0UserProfile alloc] initWithDictionary:dict];
                                                    }
                                                }];
    [dataTask resume];
}

// UIPickerViewDataSource delegate methods
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView
{
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component
{
    return self.pickerData.count;
}

- (NSString*)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component
{
    return self.pickerData[row];
}

- (void)pickerView:(UIPickerView *)pickerView didSelectRow:(NSInteger)row inComponent:(NSInteger)component
{
    self.selectedUser = [self.userList[row] valueForKey:@"user_id" ];
}

// Internal methods
- (void)createUserList:(NSDictionary *)userList {
    self.userList = [NSMutableArray arrayWithArray:userList[@"users"]];
    [self.pickerData removeAllObjects];
    for (NSDictionary *user in self.userList) {
        //A0UserProfile *userprofile = [[A0UserProfile alloc] initWithDictionary:user];
        [self.pickerData addObject:user[@"name"]];
    }
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [self.usersPickerView reloadAllComponents];
    });
}

- (void)showMessage:(NSString *)message {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Auth0" message:message preferredStyle:UIAlertControllerStyleAlert];
        
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil];
        [alertController addAction:ok];
        
        [self presentViewController:alertController animated:YES completion:nil];
    });
}

- (void)loginAsWithURL:(NSURL *)url {
    NSString *query = url.query;
    if ([query containsString:@"code="]) {
        NSRange range = [query rangeOfString:@"code="];
        NSString *code = [query substringFromIndex:range.length];
        NSString *absoluteUrl = url.absoluteString;
        range = [absoluteUrl rangeOfString:@"?"];
        NSString *redirectUrl = [absoluteUrl substringToIndex:range.location];
        [self fetchTokenWithCode:code callbackURL:redirectUrl];
        [self clearTextFields];
    }
    
}

- (void)setProfile:(A0UserProfile *)profile {
    _profile = profile;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        self.userName.text = profile.name;
        self.userId.text = profile.userId;
        self.userEmail.text = profile.email;
    });
}

- (void)clearTextFields {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.emailTextField.text = @"";
        self.passwordTextField.text = @"";
    });
    [self showMessage:@"NEW USER! Please check in section \"Current Connection\""];
}

@end

File: AKSwiftAuth0Test/AKSwiftAuth0Test/ViewController.swift
//
//  ViewController.swift
//  AKSwiftAuth0Test
//

import UIKit
import Lock

//Please use your Auth0 APIv2 token from https://auth0.com/docs/api/management/v2/tokens
let kAuth0APIv2Token = "Auth0APIv2Token"

//Please use your application data from https://manage.auth0.com/#/account/advanced
//section "Global Client Information"
let kGlobalClientId = "GlobalClientId"
let kGlobalClientSecret = "GlobalClientSecret"

//Please use your application data from https://auth0.com/docs/api/authentication
let kAuth0ClientSecret = "Auth0ClientSecret"
let kAuth0ClientId = "Auth0ClientId"
let kAuth0Domain = "Auth0Domain"

let kAuth0ConnectionType = "Username-Password-Authentication"
let kOpenURLProperty = "openURL"
let kAccessToken = "access_token"

class ViewController: UIViewController, UIPickerViewDataSource, UIPickerViewDelegate {

    @IBOutlet weak var usersPickerView: UIPickerView!
    @IBOutlet weak var emailLabel: UILabel!
    @IBOutlet weak var userIdLabel: UILabel!
    @IBOutlet weak var usernameLabel: UILabel!
    @IBOutlet weak var emailTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    
    var pickerData:NSMutableArray = []
    var profile:A0UserProfile?
    var tokens:A0Token?
    var userList:NSArray?
    var selectedUser:String?

    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.usersPickerView.dataSource = self
        self.usersPickerView.delegate = self
    }
    
    // Add observing for "openURL" property of AppDelegate
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        let appDelegate:AppDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        let options = NSKeyValueObservingOptions([.New, .Old])
        appDelegate.addObserver(self, forKeyPath:kOpenURLProperty, options:options, context:nil);        
    }
  
    override func viewDidDisappear(animated: Bool) {
        super.viewDidDisappear(animated)
        let appDelegate:AppDelegate = UIApplication.sharedApplication().delegate as! AppDelegate
        appDelegate.removeObserver(self, forKeyPath: kOpenURLProperty)
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        guard keyPath != nil else {
            super.observeValueForKeyPath(keyPath, ofObject: object, change: change, context: context)
            return
        }
        
        let oldURL:NSURL? = change![NSKeyValueChangeOldKey] as? NSURL
        if let newURL = change![NSKeyValueChangeNewKey] as? NSURL {
            if (keyPath! == kOpenURLProperty) && !(oldURL == newURL) {
                self.loginAsWithURL(newURL)
            }
        }
    }
    
    // Step 1: Login to Auth0
    @IBAction func clickLoginButton(sender: AnyObject) {
        if (self.emailTextField.text?.characters.count < 1) {
            self.showMessage("Please eneter an email");
            return;
        }
        if (self.passwordTextField.text?.characters.count < 1) {
            self.showMessage("Please eneter a password");
            return;
        }
        
        let success = { (profile: A0UserProfile?, token: A0Token?) in
            self.tokens = token
            self.profile = profile
            self.showUserProfile(profile!)
        }
        let failure = { (error: NSError) in
            self.clearUserProfile()
            self.showMessage("Oops something went wrong: \(error)");
            print("Oops something went wrong: \(error)")
        }
        
        let email = self.emailTextField.text!;
        let password = self.passwordTextField.text!;
        let client = A0Lock.sharedLock().apiClient()
        let parameters = A0AuthParameters(dictionary: [A0ParameterConnection : kAuth0ConnectionType])
        
        //Login with email and password (Auth0 database connection)
        client.loginWithUsername(email, password: password, parameters: parameters, success: success, failure: failure)
    }
    
    // Step 2: Get list of available users for impersonation
    @IBAction func clickGetUserListButton(sender: AnyObject) {
        // GET request
        // We need url "https://<Auth0 Domain>//api/v2/users?include_totals=true&include_fields=true&search_engine=v2"
        // and header "Authorization : Bearer <kAuth0APIv2Token>"
        
        let userDomain = (NSBundle.mainBundle().infoDictionary![kAuth0Domain]) as! String
        let urlString = "https://\(userDomain)/api/v2/users?include_totals=true&include_fields=true&search_engine=v2"
        let url = NSURL(string: urlString)
        if let actualUrl = url {
            let request = NSMutableURLRequest(URL: actualUrl)
            request.HTTPMethod = "GET";
            let apiToken = (NSBundle.mainBundle().infoDictionary![kAuth0APIv2Token]) as! String
            request.allHTTPHeaderFields = ["Authorization" : "Bearer \(apiToken)"]
            
            NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: {(data : NSData?, response : NSURLResponse?, error : NSError?) in
                
                // Check if data was received successfully
                if error == nil && data != nil {
                    do {
                        // Convert NSData to Dictionary where keys are of type String, and values are of any type
                        let json = try NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableContainers) as! [String:AnyObject]
                        print("\(json)")
                        self.createUserList(json)
                    } catch {
                        let dataString = String(data: data!, encoding: NSUTF8StringEncoding)
                        print("Oops something went wrong: \(dataString)")
                    }
                } else {
                    print("Oops something went wrong: \(error)")
                }
            }).resume()
        } else {
            print("Incorrect url")
        }
    }
    
    // Step 3: Get bearer token for an impersonation
    @IBAction func clickImpersonateButton(sender: AnyObject) {
        // POST request
        // We need url "https://<Auth0 Domain>/oauth/token"
        // and header "content-type": "application/json"
        
        let headers = ["content-type": "application/json"]
        let clientId = (NSBundle.mainBundle().infoDictionary![kGlobalClientId]) as! String
        let clientSecrete = (NSBundle.mainBundle().infoDictionary![kGlobalClientSecret]) as! String
        let parameters = [
            "client_id": clientId,
            "client_secret": clientSecrete,
            "grant_type": "client_credentials"
        ]
        
        var postData:NSData = NSData()
        do {
            postData = try NSJSONSerialization.dataWithJSONObject(parameters, options:[])
        } catch let error as NSError {
            print(error.localizedDescription)
            return;
        }
        
        let userDomain = (NSBundle.mainBundle().infoDictionary!["Auth0Domain"]) as! String
        let request = NSMutableURLRequest(URL: NSURL(string: "https://\(userDomain)/oauth/token")!,
                                          cachePolicy: .UseProtocolCachePolicy,
                                          timeoutInterval: 10.0)
        request.HTTPMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.HTTPBody = postData
        
        NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: { (data, response, error) -> Void in
            // Check if data was received successfully
            if error == nil && data != nil {
                do {
                    // Convert NSData to Dictionary where keys are of type String, and values are of any type
                    let json = try NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableContainers) as! [String:AnyObject]
                    print("\(json)")
                    //Impersonate user
                    self.impersonateUserWithBearerToken(json[kAccessToken] as? String)
                } catch {
                    let dataString = String(data: data!, encoding: NSUTF8StringEncoding)
                    print("Oops something went wrong: \(dataString)")
                }
            } else {
                print("Oops something went wrong: \(error)")
            }
        }).resume()
    }
    
    // Step 4: Perform the impersonation for selected user (user_id) and get a link
    func impersonateUserWithBearerToken(bearerToken:String?) {
        // POST request
        // We need url "https://<Auth0 Domain>/users/<Selected_User_id>/impersonate"
        // and header "authorization": "Bearer <Bearer_Token>"

        if let actualToken = bearerToken {
            let headers = [
                "content-type": "application/json",
                "authorization": "Bearer \(actualToken)"
            ]
            let parameters = [
                "protocol": "oauth2",
                "impersonator_id": self.profile!.userId,
                "client_id": ((NSBundle.mainBundle().infoDictionary![kAuth0ClientId]) as! String),
                "response_type": "code"
            ]

            var postData:NSData = NSData()
            do {
                postData = try NSJSONSerialization.dataWithJSONObject(parameters, options:[])
            } catch let error as NSError {
                print(error.localizedDescription)
                return;
            }
            
            let userDomain = (NSBundle.mainBundle().infoDictionary![kAuth0Domain]) as! String
            let userId = self.selectedUser!.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.alphanumericCharacterSet());

            let request = NSMutableURLRequest(URL: NSURL(string: "https://\(userDomain)/users/\(userId!)/impersonate")!,
                                              cachePolicy: .UseProtocolCachePolicy,
                                              timeoutInterval: 10.0)
            request.HTTPMethod = "POST"
            request.allHTTPHeaderFields = headers
            request.HTTPBody = postData
            
             NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: { (data, response, error) -> Void in
                // Check if data was received successfully
                if error == nil && data != nil {
                    let url = NSString(data: data!, encoding: NSUTF8StringEncoding)
                    self.openURLImpersonation(url as! String);
                } else {
                    print("Oops something went wrong: \(error)")
                }
             }).resume()
        }
    }
    
    // Step 5: Open the link for impersonation
    func openURLImpersonation(urlImpersonation:String) {
        UIApplication.sharedApplication().openURL(NSURL(string: urlImpersonation)!)
    }
    
    // Step 6: Fetch a token with the code
    func fetchTokenWithCode(code: String, callbackURL:String) {
        // POST request
        // We need url "https://<Auth0 Domain>/oauth/token"
        // and header "content-type": "application/json"
        
        let userDomain = (NSBundle.mainBundle().infoDictionary![kAuth0Domain]) as! String
        let clientId = (NSBundle.mainBundle().infoDictionary![kAuth0ClientId]) as! String
        let clientSecrete = (NSBundle.mainBundle().infoDictionary![kAuth0ClientSecret]) as! String
        
        let headers = ["content-type": "application/json"]
        let parameters = [
            "client_id": clientId,
            "client_secret": clientSecrete,
            "grant_type": "authorization_code",
            "redirect_uri": callbackURL,
            "code": code
        ]
        
        var postData:NSData = NSData()
        do {
            postData = try NSJSONSerialization.dataWithJSONObject(parameters, options:[])
        } catch let error as NSError {
            print(error.localizedDescription)
            return;
        }
        
        let request = NSMutableURLRequest(URL: NSURL(string: "https://\(userDomain)/oauth/token")!,
                                          cachePolicy: .UseProtocolCachePolicy,
                                          timeoutInterval: 10.0)
        request.HTTPMethod = "POST"
        request.allHTTPHeaderFields = headers
        request.HTTPBody = postData
        
        NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: { (data, response, error) -> Void in
            // Check if data was received successfully
            if error == nil && data != nil {
                do {
                    // Convert NSData to Dictionary where keys are of type String, and values are of any type
                    let json = try NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableContainers) as! [String:AnyObject]
                    print("\(json)")
                    //Fetch user profile
                    self.fetchUserProfileWithAccessToken(json[kAccessToken] as? String)
                } catch {
                    let dataString = String(data: data!, encoding: NSUTF8StringEncoding)
                    print("Oops something went wrong: \(dataString)")
                }
            } else {
                print("Oops something went wrong: \(error)")
            }
        }).resume()
    }
    
    // Step 7: Get user profile for a new user
    func fetchUserProfileWithAccessToken(accessToken:String?) {
        // GET request
        // We need url "https://<Auth0 Domain>/userinfo/?access_token=<ACCESS_TOKEN>"

        if let actualAccessToken = accessToken {
            let userDomain = (NSBundle.mainBundle().infoDictionary!["Auth0Domain"]) as! String
            let url = NSURL(string: "https://\(userDomain)/userinfo?access_token=\(actualAccessToken)")
            if let actualUrl = url {
                let request = NSMutableURLRequest(URL: actualUrl)
                request.HTTPMethod = "GET";
                
                NSURLSession.sharedSession().dataTaskWithRequest(request, completionHandler: {(data : NSData?, response : NSURLResponse?, error : NSError?) in
                    
                    // Check if data was received successfully
                    if error == nil && data != nil {
                        do {
                            // Convert NSData to Dictionary where keys are of type String, and values are of any type
                            let json = try NSJSONSerialization.JSONObjectWithData(data!, options: NSJSONReadingOptions.MutableContainers) as! [String:AnyObject]
                            print("\(json)")
                            self.profile = A0UserProfile(dictionary:json);
                            self.showUserProfile(self.profile!)
                        } catch {
                            let dataString = String(data: data!, encoding: NSUTF8StringEncoding)
                            print("Oops something went wrong: \(dataString)")
                        }
                    } else {
                        print("Oops something went wrong: \(error)")
                    }
                }).resume()
            } else {
                print("Incorrect url")
            }
        }
    }
    
    // UIPickerViewDataSource delegate methods
    func numberOfComponentsInPickerView(pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
        return self.pickerData.count
    }
    
    func pickerView(pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return self.pickerData[row] as? String
    }
    
    func pickerView(pickerView: UIPickerView, didSelectRow row: Int, inComponent component: Int) {
        let actualUser:NSDictionary = self.userList![row] as! NSDictionary
        self.selectedUser = actualUser["user_id"] as? String
    }

    // Internal methods
    func showMessage(message: String) {
        dispatch_async(dispatch_get_main_queue()) {
            let alert = UIAlertController(title: "Auth0", message: message, preferredStyle: UIAlertControllerStyle.Alert)
            alert.addAction(UIAlertAction(title: "OK", style: UIAlertActionStyle.Default, handler: nil))
            self.presentViewController(alert, animated: true, completion: nil)
        }
    }
    
    func showUserProfile(profile: A0UserProfile) {
        dispatch_async(dispatch_get_main_queue()) {
            self.usernameLabel.text = profile.name
            self.emailLabel.text = profile.email
            self.userIdLabel.text = profile.userId
        }
    }
    
    func clearUserProfile() {
        dispatch_async(dispatch_get_main_queue()) {
            self.usernameLabel.text = ""
            self.emailLabel.text = ""
            self.userIdLabel.text = ""
        }
    }
    
    func loginAsWithURL(url:NSURL) {
        let query = url.query!
        if query.rangeOfString("code=") != nil {
            let code = query[query.startIndex.advancedBy(5)..<query.endIndex]
            let absoluteUrl = url.absoluteString
            let redirectUrl = absoluteUrl[absoluteUrl.startIndex..<absoluteUrl.endIndex.advancedBy(-query.characters.count-1)]
            self.fetchTokenWithCode(code, callbackURL:redirectUrl)
            self.clearUserProfile();
            self.showMessage("NEW USER! Please check in section \"Current Connection\"")
        }
    }
    
    func createUserList(userList: NSDictionary) {
        self.userList = userList["users"] as? NSArray
        if let actualUserList = self.userList {
            self.pickerData.removeAllObjects()
            for user in actualUserList {
                let actualUser:NSDictionary = user as! NSDictionary
                self.pickerData.addObject(actualUser["name"]!)
            }
            dispatch_async(dispatch_get_main_queue()) {
                self.usersPickerView.reloadAllComponents()
            }
        }
    }
}
